File: net.akehurst.language/util/src/main/java/net/akehurst/language/util/SemanticAnalyserVisitorBasedAbstract.java
package net.akehurst.language.util;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import net.akehurst.holser.reflect.BetterMethodFinder;
import net.akehurst.language.api.analyser.GrammarLoader;
import net.akehurst.language.api.analyser.SemanticAnalyser;
import net.akehurst.language.api.analyser.UnableToAnalyseExeception;
import net.akehurst.language.api.sppt.SPPTBranch;
import net.akehurst.language.api.sppt.SPPTLeaf;
import net.akehurst.language.api.sppt.SPPTNode;
import net.akehurst.language.api.sppt.SharedPackedParseTree;
import net.akehurst.language.api.sppt.SharedPackedParseTreeVisitor;

public abstract class SemanticAnalyserVisitorBasedAbstract implements SemanticAnalyser, SharedPackedParseTreeVisitor<Object, Object, UnableToAnalyseExeception> {

	static Class<?>[] parameterTypes = BranchHandler.class.getMethods()[0].getParameterTypes();

	static public interface BranchHandler<T> {
		T handle(SPPTBranch target, List<SPPTBranch> children, Object arg) throws UnableToAnalyseExeception;
	}

	public SemanticAnalyserVisitorBasedAbstract() {
		this.branchHandlers = new HashMap<>();
	}

	private GrammarLoader grammarLoader;
	private final Map<String, BranchHandler> branchHandlers;

	protected void register(final String branchName, final BranchHandler handler) {
		this.branchHandlers.put(branchName, handler);
	}

	private <T> BranchHandler<T> getBranchHandler(final String branchName) throws UnableToAnalyseExeception {
		BranchHandler<T> handler = this.branchHandlers.get(branchName);
		if (null == handler) {
			try {
				final BetterMethodFinder bmf = new BetterMethodFinder(this.getClass());
				final Method m = bmf.findMethod(branchName, SemanticAnalyserVisitorBasedAbstract.parameterTypes);
				handler = (target, children, arg) -> {
					try {
						return (T) m.invoke(this, target, children, arg);
					} catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
						throw new UnableToAnalyseExeception("Error invoking method named " + branchName, e);
					}
				};
				this.register(branchName, handler);
			} catch (final NoSuchMethodException e) {
				throw new UnableToAnalyseExeception("Cannot find analyser method named " + branchName, e);
			}

		}
		return handler;
	}

	protected <T> T analyse(final SPPTBranch branch, final Object arg) throws UnableToAnalyseExeception {
		return null == branch ? null : (T) branch.accept(this, arg);
	}

	// --- ISemanticAnalyser ---
	@Override
	public <T> T analyse(final Class<T> targetType, final SharedPackedParseTree forest) throws UnableToAnalyseExeception {
		return (T) this.visit(forest, null);
	}

	@Override
	public GrammarLoader getGrammarLoader() {
		return this.grammarLoader;
	}

	@Override
	public void setGrammarLoader(final GrammarLoader value) {
		this.grammarLoader = value;
	}

	@Override
	public void clear() {
		// TODO Auto-generated method stub

	}

	// --- IParseTreeVisitor ---
	@Override
	public Object visit(final SharedPackedParseTree target, final Object arg) throws UnableToAnalyseExeception {
		final SPPTNode root = target.getRoot();
		return root.accept(this, arg);
	}

	@Override
	public Object visit(final SPPTLeaf target, final Object arg) throws UnableToAnalyseExeception {
		return target.getMatchedText();
	}

	@Override
	public Object visit(final SPPTBranch target, final Object arg) throws UnableToAnalyseExeception {
		final String branchName = target.getName();
		final BranchHandler<?> handler = this.getBranchHandler(branchName);
		if (null == handler) {
			throw new UnableToAnalyseExeception("Branch not handled in analyser " + branchName, null);
		} else {
			final List<SPPTBranch> branchChildren = target.getBranchNonSkipChildren();// .stream().map(it -> it.getIsEmpty() ? null :
																						// it).collect(Collectors.toList());
			return handler.handle(target, branchChildren, arg);
		}

	}

}

File: net.akehurst.language/processor/src/main/kotlin/processor/SemanticAnalyserVisitorBasedAbstract.kt
/**
 * Copyright (C) 2018 Dr. David H. Akehurst (http://dr.david.h.akehurst.net)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.akehurst.language.processor

import net.akehurst.language.api.analyser.GrammarLoader
import net.akehurst.language.api.sppt.SPPTBranch
import net.akehurst.language.api.sppt.SPPTLeaf
import net.akehurst.language.api.sppt.SharedPackedParseTree
import net.akehurst.language.api.sppt.SharedPackedParseTreeVisitor
import net.akehurst.language.api.sppt2ast.Sppt2AstTransformer
import net.akehurst.language.api.sppt2ast.UnableToTransformSppt2AstExeception

typealias BranchHandler<T> = (SPPTBranch, List<SPPTBranch>, Any?) -> T

abstract class Sppt2AstTransformerVisitorBasedAbstract : Sppt2AstTransformer, SharedPackedParseTreeVisitor<Any, Any?> {

    private var grammarLoader: GrammarLoader? = null
    private val branchHandlers: MutableMap<String, BranchHandler<*>> = mutableMapOf()

    protected fun <T> register(branchName: String, handler: BranchHandler<T>) {
        this.branchHandlers[branchName] = handler
    }

    private fun <T> findBranchHandler(branchName: String): BranchHandler<T> {
        var handler: BranchHandler<T>? = this.branchHandlers[branchName] as BranchHandler<T>?
        if (null == handler) {
            /*
            try {
                val m = Sppt2AstTransformerVisitorBasedAbstract::class.constructors.find {
                    it.parameters[0].type == SPPTBranch::class
                            && it.parameters[1].type == List::class
                            && it.parameters[2].type == Any::class
                }
                val handler = {target:SPPTBranch, children:List<SPPTBranch>, arg:Any? ->
                    try {
                        m?.call(this, target, children, arg) as T
                    } catch (e: Exception) {
                        throw UnableToTransformSppt2AstExeception("Error invoking method named $branchName", e)
                    }
                }
                this.register<T>(branchName, handler)
            } catch (e: Exception) {
            */
                throw UnableToTransformSppt2AstExeception("Cannot find sppt2ast method named $branchName", null)
            //}

        }
        return handler ?: throw UnableToTransformSppt2AstExeception("Cannot find sppt2ast method named $branchName", null)
    }

    protected fun <T> transform(branch: SPPTBranch?, arg: Any?): T? {
        return if (null == branch) null else branch.accept(this, arg) as T
    }

    // --- IParseTreeVisitor ---
    override fun visit(target: SharedPackedParseTree, arg: Any?): Any {
        val root = target.root
        return root.accept(this, arg)
    }

    override fun visit(target: SPPTLeaf, arg: Any?): Any {
        return target.matchedText
    }

    override fun visit(target: SPPTBranch, arg: Any?): Any {
        val branchName = target.name
        val handler = this.findBranchHandler<Any>(branchName)
        if (null == handler) {
            throw UnableToTransformSppt2AstExeception("Branch not handled in sppt2ast $branchName", null)
        } else {
            val branchChildren = target.branchNonSkipChildren// .stream().map(it -> it.getIsEmpty() ? null :
            // it).collect(Collectors.toList());
            try {
                return handler.invoke(target, branchChildren, arg)
            } catch (e: Exception) {
                throw UnableToTransformSppt2AstExeception("Exception trying to transform ${target}",e)
            }
        }

    }

}